applications/virtual-fly-brain/client/src/components/ThreeDCanvas.js

```diff
@@ -55,213 +53,57 @@ class ThreeDCanvas extends Component {
-    }
-    instances?.find( i => i.wrappedObj?.id === instance.id ) ? null : window.Instances = [...instances, instance1]
-    let that = this;
-    window.Instances.forEach( inst => {
-      inst.color = that.props.allLoadedInstances?.find( i => inst.wrappedObj.id === i.metadata.Id ).color;
-    })
-    augmentInstancesArray(window.Instances);
-  }
-  
-  getProxyInstances () {
-    return window.Instances.map(i => (
-      { ...i,
-        instancePath: i.getId(),
-        visible : this.state.mappedCanvasData?.find( cd => cd.instancePath === i.getId())?.visible,
-        color : i.color
-      }
-    ))
-  }
-
-  updateColors ( inst, mappedCanvasData) {
-    let match = mappedCanvasData?.find( m => m.instancePath === inst?.metadata?.Id )
-    let color = inst.color;
-    let colorMatch = match?.color?.b === color?.b && match?.color?.r === color?.r && match?.color?.g === color?.g;
-    if ( !colorMatch && inst?.color && match ){
-        match.color = color;
-        this.canvasRef.current.threeDEngine.updateInstances(mappedCanvasData)
-    }
-  }
-
-  newInstance (instance) {
-    this.loadInstances(instance)
-    const data = this.getProxyInstances();
-    const newData = data.map((item) => {
-      return {
-        color: item.color,
-        instancePath: item.instancePath
-      };
-    });
-    instance.setGeometryType && instance.setGeometryType('cylinders')
-
-    newData?.forEach( dat => {
-      dat.visible = data?.find( i => i.Id === data.instancePath )?.visible;
-    })
-    let mappedCanvasData = [...newData]
-    let match = mappedCanvasData?.find( m => instance.id === m.instancePath )
-    if ( match ){
-      match.visible = true;
-      this?.canvasRef?.current?.threeDEngine?.updateInstances(mappedCanvasData)
-      this.setState({ ...this.state, mappedCanvasData : mappedCanvasData})
-    }
-  }
-
-  componentDidUpdate(prevProps, prevState) {
+  componentDidUpdate(prevProps) {
     if(this.props.event.trigger !== prevProps.event.trigger){ 
-      let that = this;
-      let allLoadedInstances = this.props.allLoadedInstances;
-      const mappedCanvasData = [...this.state.mappedCanvasData];
-      const targetInstance = allLoadedInstances?.find( i => i.metadata?.Id === this.props.event.id)
-      const focusInstance = window.Instances?.find( instance => instance.wrappedObj.id === this.props.focusInstance?.Id);
+      const mappedCanvasData = [...this.props.allLoadedInstances];
+      const threeDObjects = [...this.props.threeDObjects];
 
       switch(this.props.event.action){
         case getInstancesTypes.FOCUS_INSTANCE:
-          if ( focusInstance){
-            this.canvasRef.current.threeDEngine.cameraManager.zoomTo([focusInstance])
-          } else {
-            this.canvasRef.current.defaultCameraControlsHandler("cameraHome")
-          }
           break;
-        case getInstancesTypes.CHANGE_COLOR:
-          this.updateColors(targetInstance,mappedCanvasData)
+        case getInstancesTypes.ADD_INSTANCE:{
```

Comment:
this should happen in the reducer

Reply:
@ddelpiano get3DMesh() is a redux action call that happens when the 3D Mesh is needed. Calling it from the reducer will be anti pattern. The idea behind this, is to let the Canvas request the 3D Mesh when it needs it.